snuba/query/matchers.py

```diff
@@ -0,0 +1,317 @@
+                return ret
+        return None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[ColumnExpr]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each, a Pattern can be
+    provided. If one of these Patterns is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[str]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, ColumnExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.table_name, node.table_name),
+        ):
+            if pattern is not None:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        # column_name does not have the same type as alias and table_name
+        # so this case cannot be handled in the loop above as it does not
+        # type check.
+        if self.column_name is not None:
+            partial_result = self.column_name.match(node.column_name)
+            if not partial_result:
+                return None
+            result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class Literal(Pattern[LiteralExpr]):
+    alias: Optional[Pattern[Optional[str]]] = None
+    value: Optional[Pattern[OptionalScalarType]] = None
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, LiteralExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.value, node.value),
+        ):
+            if pattern is not None:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class FunctionCall(Pattern[FunctionCallExpr]):
+    """
+    Matches a Function in the AST expression.
+    It works like the Column Pattern. if alias, function_name and function_parameters
+    are provided, they have to match, otherwise they are ignored.
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    function_name: Optional[Pattern[str]] = None
+    parameters: Optional[Tuple[Pattern[Expression], ...]] = None
+    # Specifies whether we allow optional parameters when matching.
+    # if this is False, all patterns of the function to match must match
+    # one by one. If with_optionals is True, this will allow additional
+    # parameters to exist in the function to match that are not present
+    # in this pattern.
```

Comment:
It might also help to clarify that when False, the parameter patterns list (or `Tuple`, I guess) and the actual parameter list (`Tuple`) must be the exact same length.